bson/bson_test.go

```diff
@@ -274,39 +274,125 @@ func (s *S) TestMarshalBuffer(c *C) {
 }
 
 func (s *S) TestPtrInline(c *C) {
-	cases := []struct {
-		In  interface{}
-		Out bson.M
-	}{
-		{
-			In:  inlinePtrStruct{A: 1, MStruct: &MStruct{M: 3}},
-			Out: bson.M{"a": 1, "m": 3},
-		},
-		{ // go deeper
-			In:  inlinePtrPtrStruct{B: 10, inlinePtrStruct: &inlinePtrStruct{A: 20, MStruct: &MStruct{M: 30}}},
-			Out: bson.M{"b": 10, "a": 20, "m": 30},
-		},
-		{
-			// nil embed struct
-			In:  &inlinePtrStruct{A: 3},
-			Out: bson.M{"a": 3},
-		},
-		{
-			// nil embed struct
-			In:  &inlinePtrPtrStruct{B: 5},
-			Out: bson.M{"b": 5},
-		},
+
+	// struct with inline pointer
+	{
+		in := InlineG1{G1: 4, Final: &Final{G0: 8}}
+		c.Assert(in.Final, NotNil)
+
+		data, err := bson.Marshal(in)
+		c.Assert(err, IsNil)
+
+		var out InlineG1
+		err = bson.Unmarshal(data, &out)
+		c.Assert(err, IsNil)
+		c.Assert(out.Final, NotNil)
+		c.Assert(out, DeepEquals, in)
+	}
+
+	// deeper struct with inline pointer
+	{
+		in := InlineG2{G2: 15, InlineG1: &InlineG1{G1:16, Final: &Final{G0: 23}}}
+		c.Assert(in.InlineG1, NotNil)
+		c.Assert(in.Final, NotNil)
+
+		data, err := bson.Marshal(in)
+		c.Assert(err, IsNil)
+
+		var out InlineG2
+		err = bson.Unmarshal(data, &out)
+		c.Assert(err, IsNil)
+		c.Assert(out.InlineG1, NotNil)
+		c.Assert(out.Final, NotNil)
+		c.Assert(out, DeepEquals, in)
+	}
+
+	// struct with nil inline pointer
+	{
+		in := InlineG1{G1: 42}
+		c.Assert(in.Final, IsNil)
+
+		data, err := bson.Marshal(in)
+		c.Assert(err, IsNil)
+
+		// default behaviour: no respect to nil values
+		var out InlineG1
+		err = bson.Unmarshal(data, &out)
+		c.Assert(err, IsNil)
+		c.Assert(out.Final, NotNil)
+		c.Assert(out.G1, Equals, in.G1)
+		c.Assert(out.G0, Equals, 0)
+
+		// respect to nil value
+		var outRespectNils InlineG1
+		bson.SetRespectNilValues(true)
+		err = bson.Unmarshal(data, &outRespectNils)
+		bson.SetRespectNilValues(false)
+		c.Assert(err, IsNil)
+		c.Assert(outRespectNils.Final, IsNil)
+		c.Assert(outRespectNils, DeepEquals, in)
 	}
 
-	for _, cs := range cases {
-		data, err := bson.Marshal(cs.In)
+	// deeper struct with nil inline pointer
+	{
+		in := InlineG2{G2: 108}
+		c.Assert(in.InlineG1, IsNil)
+
+		data, err := bson.Marshal(in)
+		c.Assert(err, IsNil)
+
+		// default behaviour: no respect to nil values
+		var out InlineG2
+		err = bson.Unmarshal(data, &out)
 		c.Assert(err, IsNil)
-		var dataBSON bson.M
-		err = bson.Unmarshal(data, &dataBSON)
+		c.Assert(out.InlineG1, NotNil)
+		c.Assert(out.Final, NotNil)
+		c.Assert(out.G2, Equals, in.G2)
+		c.Assert(out.G1, Equals, 0)
+		c.Assert(out.G0, Equals, 0)
+
+		// respect to nil value
+		var outRespectNils InlineG2
+		bson.SetRespectNilValues(true)
+		err = bson.Unmarshal(data, &outRespectNils)
+		bson.SetRespectNilValues(false)
 		c.Assert(err, IsNil)
+		c.Assert(outRespectNils.InlineG1, IsNil)
+		c.Assert(outRespectNils, DeepEquals, in)
+	}
 
-		c.Assert(dataBSON, DeepEquals, cs.Out)
+	// mixing empty & nil
+	// @TODO
```

Comment:
Is this just a forgotten test?

Reply:
Hm, I think it was a real TODO, not just forgotten test.
I had really turbulent few of weeks too, so I think I need to spend couple hours on this yet.

Let me look into it this weekend and then I'll push / or post here the update status

Reply:
@domodwyer 

I've debugged the issue:

It's expected behaviour in the failing test, as `g1` field is `omitempty` it is lost while `marshalling`. That's why when we unmarshal back with `respect-nil` option - we lose it.

Removing `omitempty` from `g1` field from `InlineG1` struct makes test pass. 
So either remove `omitempty`, either remove that test.